packages/slate/src/controllers/editor.js

```diff
@@ -0,0 +1,312 @@
+    }
+
+    return next()
+  }
+
+  /**
+   * Set the `onChange` handler.
+   *
+   * @param {Function} onChange
+   * @return {Editor}
+   */
+
+  setOnChange(onChange) {
+    this.onChange = onChange
+    return this
+  }
+
+  /**
+   * Set the editor's plugins.
+   *
+   * @param {Array} rawPlugins
+   * @return {Editor}
+   */
+
+  setPlugins(rawPlugins) {
+    // PERF: If they try to set the same `plugins` again, we can avoid work.
+    if (rawPlugins === this.tmp.rawPlugins) {
+      return this
+    }
+
+    // PERF: Save a reference to the "raw" plugins that were set, so that we can
+    // compare it by reference for a future set to avoid repeating work.
+    this.tmp.rawPlugins = rawPlugins
+
+    const corePlugin = CorePlugin()
+    const plugins = [corePlugin, ...rawPlugins]
+
+    class Change extends AbstractChange {}
+    const array = []
+
+    for (const plugin of plugins) {
+      const { commands, queries, schema } = plugin
+      array.push(plugin)
+
+      if (commands) {
+        const commandsPlugin = CommandsPlugin(commands)
+        array.push(commandsPlugin)
+
+        Object.keys(commands).forEach(key => {
+          Change.prototype[key] = function(...args) {
+            return this.command(key, ...args)
+          }
+        })
+      }
+
+      if (queries) {
+        const queriesPlugin = QueriesPlugin(queries)
+        array.push(queriesPlugin)
+
+        Object.keys(queries).forEach(key => {
+          Change.prototype[key] = function(...args) {
+            return this.query(key, ...args)
+          }
+        })
+      }
+
+      if (schema) {
+        const schemaPlugin = SchemaPlugin(schema)
+        array.push(schemaPlugin)
+      }
+    }
+
+    this.plugins = array
+    this.Change = Change
+    return this
+  }
+
+  /**
+   * Set `properties` on the editor.
+   *
+   * @param {Object} properties
+   * @param {Object} options
+   * @return {Editor}
+   */
+
+  setProperties(properties = {}, options) {
+    const { onChange, plugins, readOnly, value } = properties
+    if (onChange !== undefined) this.setOnChange(onChange)
+    if (plugins !== undefined) this.setPlugins(plugins)
+    if (readOnly !== undefined) this.setReadOnly(readOnly)
+    if (value !== undefined) this.setValue(value, options)
+    return this
+  }
+
+  /**
+   * Set the `readOnly` flag.
+   *
+   * @param {Boolean} readOnly
+   * @return {Editor}
+   */
+
+  setReadOnly(readOnly) {
+    this.readOnly = readOnly
+    return this
+  }
+
+  /**
+   * Set the editor's `value`.
+   *
+   * @param {Value} value
+   * @param {Options} options
+   * @return {Editor}
+   */
+
+  setValue(value, options = {}) {
+    const { normalize = true } = options
+
+    // PERF: If the plugins and value haven't changed from the last seen one, we
+    // don't have to normalize it because we know it was already normalized.
+    if (
+      normalize === false ||
+      (this.plugins === this.tmp.lastPlugins &&
+        this.value === this.tmp.lastValue)
```

Comment:
Maybe I'm not understanding, but shouldn't this be checking against the value being passed in? Like: `value === this.tmp.lastValue`?

Reply:
Yup, you're right! Good catch.